poly-commit/src/hyrax/mod.rs

```diff
@@ -0,0 +1,515 @@
+    /// than the key can support. Since the number of variables is not
+    /// considered in the prototype, this function currently simply clones the
+    /// key.
+    fn trim(
+        pp: &Self::UniversalParams,
+        _supported_degree: usize,
+        _supported_hiding_bound: usize,
+        _enforced_degree_bounds: Option<&[usize]>,
+    ) -> Result<(Self::CommitterKey, Self::VerifierKey), Self::Error> {
+        Ok((pp.clone(), pp.clone()))
+    }
+
+    /// Produces a list of commitments to the passed polynomials. Cf. the
+    /// section "Square-root commitment scheme" from the reference article.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `rng` is None, since Hyrax requires randomness in order to
+    /// commit to a polynomial
+    #[allow(unused_variables)]
+    fn commit<'a>(
+        ck: &Self::CommitterKey,
+        polynomials: impl IntoIterator<Item = &'a LabeledPolynomial<G::ScalarField, P>>,
+        rng: Option<&mut dyn RngCore>,
+    ) -> Result<
+        (
+            Vec<LabeledCommitment<Self::Commitment>>,
+            Vec<Self::CommitmentState>,
+        ),
+        Self::Error,
+    >
+    where
+        P: 'a,
+    {
+        let mut coms = Vec::new();
+        let mut states = Vec::new();
+
+        #[cfg(not(feature = "parallel"))]
+        let rng_inner = rng.expect("Committing to polynomials requires a random generator");
+
+        for l_poly in polynomials {
+            let label = l_poly.label();
+            let poly = l_poly.polynomial();
+
+            let n = poly.num_vars();
+            let dim = 1 << n / 2;
+
+            if n % 2 == 1 {
+                // Only polynomials with an even number of variables are
+                // supported in this implementation
+                return Err(Error::InvalidNumberOfVariables);
+            }
+
+            if n > ck.com_key.len() {
+                return Err(Error::InvalidNumberOfVariables);
+            }
+
+            let m = flat_to_matrix_column_major(&poly.to_evaluations(), dim, dim);
+
+            // Commiting to the matrix with one multi-commitment per row
+            let (row_coms, com_rands): (Vec<_>, Vec<_>) = cfg_iter!(m)
+                .map(|row| {
+                    #[cfg(not(feature = "parallel"))]
+                    let r = G::ScalarField::rand(rng_inner);
+                    #[cfg(feature = "parallel")]
+                    let r = G::ScalarField::rand(&mut rand::thread_rng());
+                    let c = (Self::pedersen_commit(&ck.com_key, row) + ck.h * r).into();
+                    (c, r)
+                })
+                .unzip();
+
+            let com = HyraxCommitment { row_coms };
+            let l_comm = LabeledCommitment::new(label.to_string(), com, Some(1));
+
+            coms.push(l_comm);
+            states.push(HyraxCommitmentState {
+                randomness: com_rands,
+                mat: Matrix::new_from_rows(m),
```

Comment:
I am confused about this @Antonio95 . Here, we build a row-major matrix from `m`. But `m` is a column-major representation of evaluations. So which is what? Moreover, it is unclear to me why we need `m` to be column-major in the first place.

Reply:
Hi Hossein! The method `flat_to_matrix_column_major` converts a flat list L into a matrix M interpreting the former as a column-first flattening of the latter. But, independently from that, there is the question of how that M is returned, i.e. how it is represented. As you can see in the [function documentation](https://github.com/HungryCatsStudio/poly-commit/blob/main/src/hyrax/utils.rs#L7), M is returned as a list of rows - and that is precisely the representation that the constructor `new_from_rows` expects.

I feel like the example in the documentation is clearer than the explanation therein - if you think it is necessary, we can fix the latter.

Let me know if that makes sense.

Reply:
Hi! Thanks for the explanation!
I think the issue is that `flat_to_matrix_column_major` returns vectors that are the columns of the matrix, but `new_from_rows` crafts a new matrix from the vectors of the rows. This could be a problem since we have matrix multiplication in
https://github.com/HungryCatsStudio/poly-commit/blob/ac4a14c5e7a3486d845229e0ef529ed64457648e/poly-commit/src/hyrax/mod.rs#L354
It seems to me that by this process, `t` in the line above is the transpose of `m`.


Also, I think it would be better if we do not use columns major for performance reasons. `flat_to_matrix_column_major` is a lot of work just for representing the matrix from row-major to column-major.

Reply:
Hey Hossein! As I mentioned in the first paragraph of my previous message, `flat_to_matrix_column_major` returns a list whose sublists are the *rows* of the matrix. Consider the example from the documentation:

![Screenshot 2024-01-24 at 08 52 25](https://github.com/arkworks-rs/poly-commit/assets/5489993/8ab85cda-d48c-4984-b871-82dd5d2afe35)

The "column-major" term in the function name only refers to how the flat input list is interpreted, not to the representation in which the matrix is returned. In particular,
- At no in the execution of `commit`/`open` do these functions construct a list whose sublists are columns of a matrix.
- At no point does the transpose of the polynomial coefficient matrix occur.

As an additional hint that things are done correctly (of course, insufficient on its own), note the commitment, opening and verification methods all work correctly with each other. If you transpose the matrix in either `commit` or `open`, the process will break (as expected).

Reply:
Hi Antonio!
Thanks for the explanation! The source of confusion was that I thought we should interpret the evaluation table as a row-major matrix rather than a column-major (as in Brakedown).

Reply:
I agree it can be a bit confusing! Is it clear now, then? Are you happy to leave that part of the code as it is?